lib/AST/GenericSignatureBuilder.cpp

```diff
@@ -1349,6 +1317,12 @@ bool GenericSignatureBuilder::addSameTypeRequirementBetweenArchetypes(
     T1->ConcreteTypeSource = T2->ConcreteTypeSource;
   }
 
+  // Decide which potential archetype is to be considered the representative.
```

Comment:
`compareDependentTypes` checks for a concrete type in the path of the potential archetype (which includes the current PA), so we need either both or neither to have a concrete type so that we effectively skip that check.